File: cpp/cpp_100-999/388.cc
// C++
// leetcode 388
// https://leetcode-cn.com/problems/longest-absolute-file-path/

class Solution {
public:
    int maxx(int a, int b) {
        return (a > b ? a : b);
    }

    int lengthLongestPath(string input) {
        int res = 0;
        stack<string> stk;
        int cur_size = 0;
        int n = input.size();
        for(int i = 0, j = 0; i < n;) {
            int tab_num = 0;
            bool isFile = false;
            while(i < n && input[i] == '\t') {
                tab_num++;
                i++;
            }
            j = i;
            while(j < n && input[j] != '\n') {
                if(input[j] == '.')
                    isFile = true;
                j++;
            }
            // cout<<i<<" :: "<<j<<endl;
            // 此时i为路径头，j为尾巴后面一位
            while(stk.size() > tab_num) {
                cur_size -= stk.top().size();
                stk.pop();
            }
            string tmp = input.substr(i, j-i);
            // cout<<tmp<<" size: "<<tmp.size()<<endl;
            // cout<<"now depth : "<<stk.size()<<" cur_size : "<<cur_size<<endl<<"-----"<<endl;
            stk.push(tmp);
            cur_size += j-i;
            i = j+1;
            if(isFile)
                res = maxx(res, cur_size + stk.size() - 1);
        }
        return res;
    }
};

File: Go/Go_100-999/388.go
// Go
// leetcode 388
// https://leetcode.cn/problems/longest-absolute-file-path/

func lengthLongestPath(s string) (res int) {
	n := len(s)
	var level int
	isFile := false
	stk := make([]int, 0, 4)
	for i := 0; i < n; {
		level, isFile = 0, false

		for i < n && s[i] == '\t' { // 每个 \t 多加一个层级
			level++
			i++
		}

		j := i
		for j < n && s[j] != '\n' {
			if s[j] == '.' { // . 判断文件
				isFile = true
			}
			j++
		}

		stk = stk[0:level]
		stk = append(stk, j-i)

		if isFile { // 是文件则可以刷一次 res
			tmp := max(len(stk)-1, 0) // slash数量
			for _, v := range stk {
				tmp += v
			}
			res = max(res, tmp)
		}
		i = j + 1
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}